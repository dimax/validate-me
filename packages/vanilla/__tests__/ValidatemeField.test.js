import flushPromises from 'flush-promises';

import vanillaConnector from '../';

const proxyHandler = () => (target, prop, val) => {
  target[prop] = val;

  return true;
};
const getInput = f => f;

describe('vanilla/ValidatemeField', () => {
  let form;

  beforeEach(() => {
    form = {
      setField: jest.fn(),
    };
  });

  test('Configures a text input and updates its value', async () => {
    const input = document.createElement('input');

    input.name = 'text';
    input.pattern = '\\D+';
    input.required = true;

    const connectMe = vanillaConnector(form, proxyHandler);
    const field = connectMe(input, getInput, [['len', '1', '5']]);

    await flushPromises();

    input.value = 'a';
    input.dispatchEvent(new Event('input'));

    await flushPromises();

    expect(field).toMatchSnapshot();
    expect(form.setField).toHaveBeenCalled();
  });

  test('Configures a number field and sets a value bigger than its max', async () => {
    const input = document.createElement('input');

    input.name = 'number';
    input.type = 'number';
    input.min = '1';
    input.max = '5';

    const connectMe = vanillaConnector(form, proxyHandler);
    const field = connectMe(input, getInput);

    await flushPromises();

    input.dispatchEvent(new Event('blur'));
    expect(field).toMatchSnapshot();

    input.value = '15';
    input.dispatchEvent(new Event('input'));

    expect(field).toMatchSnapshot();
  });

  test('Configures a checkbox field and selects it', async () => {
    const input = document.createElement('input');

    input.name = 'checkbox';
    input.type = 'checkbox';
    input.required = true;

    const connectMe = vanillaConnector(form, proxyHandler);
    const field = connectMe(input, getInput);

    await flushPromises();

    input.dispatchEvent(new Event('blur'));
    expect(field).toMatchSnapshot();

    input.checked = true;
    input.dispatchEvent(new Event('input'));

    expect(field).toMatchSnapshot();
  });

  test('Configures a radio list and selects one option', async () => {
    const options = document.createElement('div');

    options.className = 'options';
    options.dataset.name = 'ide';
    options.dataset.type = 'radio';

    const connectMe = vanillaConnector(form, proxyHandler);
    const field = connectMe(options, getInput, undefined, {
      vscode: 'VSCode',
      atom: 'Atom',
    });
    const checkbox = options.querySelector('input');

    checkbox.checked = true;
    checkbox.dispatchEvent(new Event('input'));

    expect(field).toMatchSnapshot();
  });

  test('Configures a checkbox list and checks every option', async () => {
    const options = document.createElement('div');

    options.className = 'options';
    options.dataset.name = 'ides';
    options.dataset.type = 'checkbox';

    const connectMe = vanillaConnector(form, proxyHandler);
    const field = connectMe(options, getInput, undefined, {
      vscode: 'VSCode',
      atom: 'Atom',
    });
    const checkboxes = options.querySelectorAll('input');

    checkboxes.forEach(checkbox => {
      checkbox.dispatchEvent(new Event('blur'));
      checkbox.checked = true;
      checkbox.dispatchEvent(new Event('input'));
    });

    expect(field).toMatchSnapshot();
  });

  test('Rehidrates an input with invalid data and touches it', async () => {
    const input = document.createElement('input');

    input.name = 'text';
    input.pattern = '\\D+';
    input.value = 'a2 wsa2';
    input.required = true;

    const connectMe = vanillaConnector(form, proxyHandler);
    const field = connectMe(input, getInput);

    await flushPromises();

    input.dispatchEvent(new Event('blur'));

    expect(field).toMatchSnapshot();
  });
});
