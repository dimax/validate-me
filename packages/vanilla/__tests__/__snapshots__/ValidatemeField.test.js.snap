// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vanilla/ValidatemeField Configures a checkbox field and selects it 1`] = `
ValidatemeField {
  "error": "The field is required.",
  "loading": false,
  "name": "checkbox",
  "required": true,
  "rules": Array [
    Object {
      "args": Array [],
      "name": "required",
      "run": [Function],
    },
  ],
  "touched": true,
  "value": "",
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Configures a checkbox field and selects it 2`] = `
ValidatemeField {
  "error": "",
  "loading": false,
  "name": "checkbox",
  "required": true,
  "rules": Array [
    Object {
      "args": Array [],
      "name": "required",
      "run": [Function],
    },
  ],
  "touched": true,
  "value": true,
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Configures a checkbox list and checks every option 1`] = `
ValidatemeField {
  "error": "",
  "name": "ides",
  "required": undefined,
  "rules": Array [],
  "touched": true,
  "value": Object {
    "atom": true,
    "vscode": true,
  },
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Configures a number field and sets a value bigger than its max 1`] = `
ValidatemeField {
  "error": "",
  "loading": false,
  "name": "number",
  "required": false,
  "rules": Array [
    Object {
      "args": Array [],
      "name": "number",
      "run": [Function],
    },
    Object {
      "args": Array [
        "1",
      ],
      "name": "min",
      "run": [Function],
    },
    Object {
      "args": Array [
        "5",
      ],
      "name": "max",
      "run": [Function],
    },
  ],
  "touched": true,
  "value": "",
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Configures a number field and sets a value bigger than its max 2`] = `
ValidatemeField {
  "error": "It must be less or equal than \\"5\\".",
  "loading": false,
  "name": "number",
  "required": false,
  "rules": Array [
    Object {
      "args": Array [],
      "name": "number",
      "run": [Function],
    },
    Object {
      "args": Array [
        "1",
      ],
      "name": "min",
      "run": [Function],
    },
    Object {
      "args": Array [
        "5",
      ],
      "name": "max",
      "run": [Function],
    },
  ],
  "touched": true,
  "value": "15",
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Configures a radio list and selects one option 1`] = `
ValidatemeField {
  "error": "",
  "name": "ide",
  "required": undefined,
  "rules": Array [],
  "value": "vscode",
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Configures a text input and updates its value 1`] = `
ValidatemeField {
  "error": "",
  "loading": false,
  "name": "text",
  "required": true,
  "rules": Array [
    Object {
      "args": Array [],
      "name": "required",
      "run": [Function],
    },
    Object {
      "args": Array [
        "\\\\D+",
      ],
      "name": "pattern",
      "run": [Function],
    },
    Object {
      "args": Array [
        "1",
        "5",
      ],
      "name": "len",
      "run": [Function],
    },
  ],
  "value": "a",
  "warning": "",
}
`;

exports[`vanilla/ValidatemeField Rehidrates an input with invalid data and touches it 1`] = `
ValidatemeField {
  "error": "Must match the following pattern: \\\\D+",
  "loading": false,
  "name": "text",
  "required": true,
  "rules": Array [
    Object {
      "args": Array [],
      "name": "required",
      "run": [Function],
    },
    Object {
      "args": Array [
        "\\\\D+",
      ],
      "name": "pattern",
      "run": [Function],
    },
  ],
  "touched": true,
  "value": "a2 wsa2",
  "warning": "",
}
`;
